src/bin/solang.rs

```diff
@@ -416,33 +431,87 @@ fn compile(matches: &ArgMatches) {
 
     let mut errors = false;
 
-    for filename in matches.get_many::<OsString>("INPUT").unwrap() {
-        match process_file(filename, &mut resolver, target, matches, &mut json, &opt) {
-            Ok(ns) => namespaces.push(ns),
-            Err(_) => {
-                errors = true;
-            }
+    // build a map of requested contract names, and a flag specifying whether it was found or not
+    let mut contract_names: HashSet<&str> = HashSet::new();
+
+    if let Some(values) = matches.get_many::<String>("CONTRACT") {
+        for v in values {
+            contract_names.insert(v.as_str());
         }
     }
 
-    if let Some("ast-dot") = matches.get_one::<String>("EMIT").map(|v| v.as_str()) {
-        exit(0);
+    for filename in matches.get_many::<OsString>("INPUT").unwrap() {
+        // TODO: this could be parallelized using e.g. rayon
+        let ns = process_file(filename, &mut resolver, target, matches, &opt);
+
+        namespaces.push((ns, filename));
     }
 
+    let mut json_contracts = HashMap::new();
+
     let std_json = *matches.get_one("STD-JSON").unwrap();
 
-    if errors {
+    for (ns, _) in &namespaces {
         if std_json {
-            println!("{}", serde_json::to_string(&json).unwrap());
-            exit(0);
+            let mut out = ns.diagnostics_as_json(&resolver);
+            json.errors.append(&mut out);
         } else {
-            eprintln!("error: not all contracts are valid");
-            exit(1);
+            ns.print_diagnostics(&resolver, verbose);
+        }
+
+        if ns.diagnostics.any_errors() {
+            errors = true;
+        }
+    }
+
+    if let Some("ast-dot") = matches.get_one::<String>("EMIT").map(|v| v.as_str()) {
+        exit(0);
+    }
+
+    // ensure we have any contracts
+    if !errors && !namespaces.iter().any(|(ns, _)| !ns.contracts.is_empty()) {
+        eprintln!("error: no contacts found");
+        errors = true;
+    }
+
+    // ensure we all the requested contracts
+    let not_found: Vec<_> = contract_names
+        .iter()
+        .filter(|name| {
+            !namespaces
+                .iter()
+                .flat_map(|(ns, _)| ns.contracts.iter())
+                .any(|contract| **name == contract.name)
```

Comment:
This freely mixes contracts from different files. What if I compile multiple files with oversecting contract names, but only want some contract that from a specific file. If this is fine or something you intended, this should go into the CLI help. Otherwise, I think the correct algorithm to do this would be to build a HashSet where we insert keys as follows:
- `"filename=contractname"` 
- `"contractname"

Then the user can do `--contracts foo,bar.sol=bar` if he want's the `foo` contract from every namespace but the `bar` contract only from `bar.sol`

Reply:
If the user has multiple files with the same contract name, then that is just a recipe for total disaster. How will this work in a dev environment like vscode?

I think the right solution is to detect this problem and error about it.

Reply:
> If the user has multiple files with the same contract name, then that is just a recipe for total disaster

I fully agree with you on this.

> How will this work in a dev environment like vscode?

What do you mean with this?

> I think the right solution is to detect this problem and error about it.

Do you mean in general, or just when the `--contract` flag is set? Here, `solc` is happily compiling the exact same contract with the same name but in different files:

```bash
solc --bin /tmp/*.sol 2>/dev/null

======= /tmp/a.sol:A =======
Binary:
6080604052348015600f57600080fd5b50606d80601d6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c80630dbe671f14602d575b600080fd5b60336035565b005b56fea2646970667358221220a816946390647c30b12c5d3b89636422c282d0132869c61c082513a7dfa981fa64736f6c63430008120033

======= /tmp/b.sol:A =======
Binary:
6080604052348015600f57600080fd5b50606d80601d6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c80630dbe671f14602d575b600080fd5b60336035565b005b56fea2646970667358221220a433fe66d6602c3827ae7e0060d4539266d520fd0f1a61cc952ab89aabf7d1bf64736f6c63430008120033
```

I think if we have to rely on contract names being unique per file name, we have a problem.

Reply:
> > How will this work in a dev environment like vscode?
> 
> What do you mean with this?

I was thinking of anchor because that's what I've been working on. So you have a bunch of solidity files on one hand, 
and then you have some client code/test code in written in typescript. Where do you specify that you want contract `A` from file X.sol but not Y.sol? Would it not be better if the environment said "error: contract A defined in X.sol and Y.sol".
 
> > I think the right solution is to detect this problem and error about it.
> 
> Do you mean in general, or just when the `--contract` flag is set?

I was thinking in general. This can always be a problem. Having said that, I don't know what can be done.

> Here, `solc` is happily compiling the exact same contract with the same name but in different files:
> 
> ```shell
> solc --bin /tmp/*.sol 2>/dev/null
> 
> ======= /tmp/a.sol:A =======
> Binary:
> 6080604052348015600f57600080fd5b50606d80601d6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c80630dbe671f14602d575b600080fd5b60336035565b005b56fea2646970667358221220a816946390647c30b12c5d3b89636422c282d0132869c61c082513a7dfa981fa64736f6c63430008120033
> 
> ======= /tmp/b.sol:A =======
> Binary:
> 6080604052348015600f57600080fd5b50606d80601d6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c80630dbe671f14602d575b600080fd5b60336035565b005b56fea2646970667358221220a433fe66d6602c3827ae7e0060d4539266d520fd0f1a61cc952ab89aabf7d1bf64736f6c63430008120033
> ```
> 
> I think if we have to rely on contract names being unique per file name, we have a problem. I'm not aware on any reason why the file names should affect compilation in any way. We could make an exception for the `--contract` flag but this would be really odd.

I don't know what to do here, if solc permits this. We could do some semantic comparison of the contracts, but this is hard to do because 
- the locs will definitely be different 
- the struct definitions etc will have different numbers because they are in different namespaces 
- we will have to recurse down any base contracts/interfaces used/libraries used, globals used

As a cop-out, I'd like to point out this is not a new problem for this PR :sweat_smile:

Reply:
I see, didn't even have the front end tooling in mind. Looks like we have uncovered something here requiring more thought. Let's move on and turn this into a documentation problem for now.

> As a cop-out, I'd like to point out this is not a new problem for this PR sweat_smile

:100:  But I'd appreciate if we could point this out in the CLI help for `--contract`

Reply:
>  But I'd appreciate if we could point this out in the CLI help for `--contract`

Absolutely right, that belongs in the docs. I've added a warning.